src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -0,0 +1,183 @@
+#include <boost/algorithm/string/predicate.hpp>
+#include <chrono>
+#include <fmt/core.h>
+#include <memory>
+#include <thread>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "Common/Exception.h"
+#include "Common/logger_useful.h"
+#include "Checkpoint/CheckpointContext.h"
+#include "Checkpoint/CheckpointContextFwd.h"
+#include "Core/Block.h"
+#include "Formats/FormatFactory.h"
+#include "Interpreters/Context.h"
+#include "KafkaLog/KafkaWALCommon.h"
+#include "KafkaSink.h"
+#include "Processors/Chunk.h"
+#include "Processors/Formats/IOutputFormat.h"
+#include "Processors/ProcessorID.h"
+#include "Storages/ExternalStream/ExternalStreamTypes.h"
+#include "Storages/ExternalStream/Kafka/Kafka.h"
+#include "Storages/ExternalStream/Kafka/WriteBufferFromKafka.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int MISSING_ACKNOWLEDGEMENT;
+extern const int INVALID_CONFIG_PARAMETER;
+}
+
+KafkaSink::KafkaSink(const Kafka * kafka, const Block & header, ContextPtr context, const Poco::Logger * logger)
+    : SinkToStorage(header, ProcessorID::ExternalTableDataSinkID), log(logger), producer(nullptr, nullptr)
+{
+    is_streaming = true;
+
+    /// default values
+    std::vector<std::pair<std::string, std::string>> producer_params{
+        std::make_pair("enable.idempotence", "true"),
+        std::make_pair("message.timeout.ms", "0" /* infinite */),
+    };
+
+    static const std::unordered_set<std::string> allowed_properties{
+        "enable.idempotence",
+        "message.timeout.ms",
+        "queue.buffering.max.messages",
+        "queue.buffering.max.kbytes",
+        "queue.buffering.max.ms",
+        "message.send.max.retries",
+        "retries",
+        "retry.backoff.ms",
+        "retry.backoff.max.ms",
+        "batch.num.messages",
+        "batch.size",
+        "compression.codec",
+        "compression.type",
+        "compression.level",
+    };
+
+    /// customization, overrides default values
+    for (const auto & pair : kafka->properties())
+    {
+        if (allowed_properties.contains(pair.first))
+        {
+            producer_params.emplace_back(pair.first, pair.second);
+            continue;
+        }
+        throw Exception("Unsupported property: " + pair.first, ErrorCodes::INVALID_CONFIG_PARAMETER);
```

Comment:
Let's stick to this format when throw

```
throw Exception(error_code, "..... format string....",  ... format args ...); 
```

`format string` is something using the format of `std::format` or `fmt::format`

Reply:
```
throw Exception(ErrorCodes::INVALID_CONFIG_PARAMETER, "Unsupported property: {}={}" + pair.first, p.second);
```